### PR TITLE
Support INTERSECT

### PR DESCRIPTION
**Pull request by anusudarsan**
Implement INTERSECT sql as a query rewrite. Added a new optimization
rule for rewriting an INTERSECT query to UNION ALL. The rewrite for:

`SELECT a from foo INTERSECT SELECT x FROM bar`
is:

`
SELECT a
  FROM
  (SELECT a,
    COUNT(foo_marker) AS foo_cnt,
    COUNT(bar_marker) AS bar_cnt
        FROM
           (
            SELECT a, true as foo_marker, null as bar_marker FROM foo
            UNION ALL
            SELECT x, null as foo_marker, true as bar_marker FROM bar
           ) T1
GROUP BY a) T2
WHERE foo_cnt >= 1 AND bar_cnt >= 1;
`
This also involves related refactoring of extracting a SetOperation for
UNION, INTERSECT and EXCEPT.

### CODE REVIEW

**Comment by anusudarsan** [comment]
@cawallin @rschlussel @petroav Can you take a look please?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -1197,17 +1204,17 @@ protected RelationType visitUnion(Union node, AnalysisContext context)
             if (outputFieldSize != descFieldSize) {
                 throw new SemanticException(MISMATCHED_SET_COLUMN_TYPES,
                         node,
-                        "union query has different number of fields: %d, %d",
-                        outputFieldSize, descFieldSize);
+                        "%s query has different number of fields: %d, %d",
+                        node.getSetOpType().name(), outputFieldSize, descFieldSize);
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Extract the set operation type name once.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            ImmutableMap.Builder<Symbol, Expression> outputProjections = ImmutableMap.builder();
+
+            for (int i = 0; i < numberOfSources; i++) {
+                ProjectNode newProjectNode = createProjectNode(node, sources, i);
+                newProjectNodes.add(newProjectNode);
+
+                List<Symbol> projectOutputSymbols = newProjectNode.getOutputSymbols();
+                if (unionOutputSymbols == null) {
+                    // Use the first Relation to derive output symbol names
+                    ImmutableList.Builder<Symbol> outputSymbolBuilder = ImmutableList.builder();
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java]: Method is too long, may be be break it down by extracting a smaller method for getting this list?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        requireNonNull(symbolAllocator, "symbolAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator, symbolAllocator), plan);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Integer>
+    {
+        private static final String INTERSECT_MARKER = "intersect_marker";
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:97]: What if the table has a column named intersect_marker? Or if it is in the SELECT list of original query?

Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:97]: So, then the symbolAllocator will generate a new unique id intersect_marker_<id> for the new ones added here. This string is just a "hint" for what the column name should be.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SetFlatteningOptimizer.java [hunk 4]
```diff
@@ -85,6 +86,34 @@ public PlanNode visitUnion(UnionNode node, RewriteContext<Boolean> context)
...
+        {
+            ImmutableList.Builder<PlanNode> flattenedSources = ImmutableList.builder();
+            ImmutableListMultimap.Builder<Symbol, Symbol> flattenedSymbolMap = ImmutableListMultimap.builder();
+            for (int i = 0; i < node.getSources().size(); i++) {
+                PlanNode subplan = node.getSources().get(i);
+                PlanNode rewrittenSource = context.rewrite(subplan, context.get());
+
+                if (rewrittenSource instanceof IntersectNode) {
+                    // Absorb source's subplans if it is also a IntersectNode
+                    IntersectNode rewrittenUnion = (IntersectNode) rewrittenSource;
```
Comment (rschlussel-zz) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SetFlatteningOptimizer.java]: should be called rewrittenIntersect

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SetFlatteningOptimizer.java [hunk 5]
```diff
@@ -85,6 +86,34 @@ public PlanNode visitUnion(UnionNode node, RewriteContext<Boolean> context)
         }
 
         @Override
+        public PlanNode visitIntersect(IntersectNode node, RewriteContext<Boolean> context)
```
Comment (rschlussel-zz) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SetFlatteningOptimizer.java]: This looks exactly the same as visitUnion except that you check that the source is an Intersect instead of union.  Is there a way to combine these into a common SetOperation method so we don't have duplicated code?

Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SetFlatteningOptimizer.java]: hmm..not sure how. This optimizer is for combining multiple Set operations into one. For eg: `a interesect b intersect c`, will be converted to one intersect node with sources a,b, c. But in case of `a union b intersect c ` we do not want to combine it into one setOperation node.

**Comment by anusudarsan** [comment]
updated.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanSanityChecker.java [hunk 6]
```diff
@@ -499,9 +501,14 @@ public Void visitTableFinish(TableFinishNode node, Void context)
...
         {
+            return visitSetOperation(node, context);
+        }
+
+        private Void visitSetOperation(SetOperationNode node, Void context)
+        {
             for (int i = 0; i < node.getSources().size(); i++) {
                 PlanNode subplan = node.getSources().get(i);
-                checkDependencies(subplan.getOutputSymbols(), node.sourceOutputLayout(i), "UNION subplan must provide all of the necessary symbols");
+                checkDependencies(subplan.getOutputSymbols(), node.sourceOutputLayout(i), "%s subplan must provide all of the necessary symbols", node.getSetOperationType().name());
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanSanityChecker.java]: do you need format() here or does checkDependencies fill in the %s for you?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 7]
```diff
@@ -613,62 +616,26 @@ protected RelationPlan visitUnion(Union node, Void context)
...
-            }
-        }
-
-        Optional<Symbol> outputSampleWeight = Optional.empty();
-        for (RelationPlan relationPlan : subPlans) {
-            if (hasSampleWeight && !relationPlan.getSampleWeight().isPresent()) {
-                relationPlan = addConstantSampleWeight(relationPlan);
-            }
+        SetOperationPlanner setOperationPlanner = new SetOperationPlanner(node, context).process();
+        ImmutableListMultimap.Builder<Symbol, Symbol> symbolMapping = setOperationPlanner.getSymbolMapping();
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: do the .build() step here, so you don't need to build it multiple times below

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 8]
```diff
@@ -780,4 +747,89 @@ private PlanNode distinct(PlanNode node)
                 1.0,
                 Optional.empty());
     }
+
+    private class SetOperationPlanner
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java:752]: Maybe this should be in a new file? This file is pretty huge already.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 9]
```diff
@@ -780,4 +747,89 @@ private PlanNode distinct(PlanNode node)
...
+        private ImmutableListMultimap.Builder<Symbol, Symbol> symbolMapping;
+        private Optional<Symbol> outputSampleWeight;
+
+        public SetOperationPlanner(SetOperation node, Void context)
+        {
+            this.node = node;
+            this.context = context;
+        }
+
+        public ImmutableList.Builder<PlanNode> getSources()
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Return sources.build() instead

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 10]
```diff
@@ -780,4 +747,89 @@ private PlanNode distinct(PlanNode node)
...
+        {
+            return sources;
+        }
+
+        public ImmutableListMultimap.Builder<Symbol, Symbol> getSymbolMapping()
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: symbolMapping.build() instead

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+ * (
+ * SELECT a, true as foo_marker, null as bar_marker FROM foo
+ * UNION ALL
+ * SELECT x, null as foo_marker, true as bar_marker FROM bar
+ * ) T1
+ * GROUP BY a) T2
+ * WHERE foo_cnt >= 1 AND bar_cnt >= 1;
+ */
+public class ImplementIntersectAsUnion
+        extends PlanOptimizer
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:80]: Hmm. This step isn't so much a "PlanOptimizer" step as much as a "we don't have an INTERSECT operator so we're going to map it to UNION" step. I'm not so familiar with the Presto planner/optimizer, is this plan optimization phase the part where logical operators are mapped to physical operators or should this go someplace else?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        private ProjectNode createProjectNode(IntersectNode intersectNode, List<PlanNode> sources, int sourceIndex)
+        {
+            ImmutableMap.Builder<Symbol, Expression> projectionAssignments = ImmutableMap.builder();
+            Map<Symbol, QualifiedNameReference> sourceSymbolMap = intersectNode.sourceSymbolMap(sourceIndex);
+
+            // add existing intersect symbols to projection
+            for (Symbol intersectSymbol : sourceSymbolMap.keySet()) {
+                Symbol symbol = allocateNewSymbolFrom(intersectSymbol);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:208]: can you just copy these or do you have to make new ones?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+            }
+            return outputSymbolBuilder.build();
+        }
+
+        private Symbol allocateNewSymbolFrom(Symbol symbol)
+        {
+            return allocateNewSymbolFrom(symbol.getName(), symbolAllocator.getTypes().get(symbol));
+        }
+
+        private AggregationNode createAggregationNode(IntersectNode node, List<Symbol> outputSymbols, UnionNode sourceNode, ImmutableMap.Builder<Symbol, FunctionCall> aggregationsBuilder, ImmutableMap.Builder<Symbol, Expression> outputProjections)
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java]: can you call this something like createAggregationNodeForIntersect()? I was confused when first reading through because I didn't see the group by things added above.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+            List<Expression> markers = new ArrayList<>(Collections.nCopies(sources.size(), new Cast(new NullLiteral(), StandardTypes.BOOLEAN)));
+            markers.set(sourceIndex, TRUE_LITERAL);
+            for (Expression markerExpression : markers) {
+                projectionAssignments.put(allocateNewSymbolFrom(INTERSECT_MARKER, BOOLEAN), markerExpression);
+            }
+
+            return new ProjectNode(idAllocator.getNextId(), sources.get(sourceIndex), projectionAssignments.build());
+        }
+
+        private FilterNode createFilterNode(AggregationNode sourceNode, Map<Symbol, FunctionCall> aggregations)
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java]: can you call this createFilterNodeForIntersect as well?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanSanityChecker.java (hunk 6 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanSanityChecker.java]: checkDependencies handles it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 11 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:80]: I think it would be the best to have it here, because 1) there is no UNION physical operator as such. Presto uses ExchangeOperator (GATHER mode) to perform union 2) These ExchangeNodes are added appropriately for the UNION operators in the PlanOptimizer (a new rule called AddExchanges here). So it will be the best to add this before that rule, so that the physical planning happens accordingly for intersect queries. and 3) AddExchanges also does some sort of distributed planning for UNION queries, which we may want to take advantage of.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 9 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 8 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java:752]: leaving it as-is, since this uses some relationPlanner methods which cannot be made static

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 11 above)
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:80]: OK, makes sense, thanks for the explanation

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 12 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:208]: maybe we can reuse, I just allocated new ones for any new nodes I create so that we have unique symbols (and no conflicts). Let me look into this a bit and check if we can reuse

**Comment by anusudarsan** [comment]
addressed comments

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 15]
```diff
@@ -780,4 +747,89 @@ private PlanNode distinct(PlanNode node)
                 1.0,
                 Optional.empty());
     }
+
+    private class SetOperationPlanner
+    {
+        private final SetOperation node;
+        private final Void context;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Void cannot be instantiated. No need to store it (null) as field.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 16]
```diff
@@ -780,4 +747,89 @@ private PlanNode distinct(PlanNode node)
...
+        private ImmutableList.Builder<PlanNode> sources;
+        private ImmutableListMultimap.Builder<Symbol, Symbol> symbolMapping;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: why do you store builders? Try to have List and ListMultimap here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+ * COUNT(foo_marker) AS foo_cnt,
+ * COUNT(bar_marker) AS bar_cnt
+ * FROM
+ * (
+ * SELECT a, true as foo_marker, null as bar_marker FROM foo
+ * UNION ALL
+ * SELECT x, null as foo_marker, true as bar_marker FROM bar
+ * ) T1
+ * GROUP BY a) T2
+ * WHERE foo_cnt >= 1 AND bar_cnt >= 1;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:77]: How INTERSECT is working when you have multiple rows with the save value for column `a` in table `foo` and one matching row in bar. Is it going to return duplicates of `a` values? If so then this rewrites won't work as recourds are grouped by `a`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+            }
+
+            UnionNode unionNode = new UnionNode(node.getId(), newProjectNodes.build(), symbolMappings.build(), ImmutableList.copyOf(symbolMappings.build().keySet()));
+            AggregationNode aggregationNode = createAggregationNodeForIntersect(node, unionOutputSymbols, unionNode, aggregationsBuilder, outputProjections);
+            FilterNode filterNode = createFilterNodeForIntersect(aggregationNode, aggregationsBuilder.build());
+
+            return new ProjectNode(idAllocator.getNextId(), filterNode, outputProjections.build());
+        }
+
+        private List<Symbol> getUnionOutputSymbolsFrom(List<Symbol> sourceSymbols)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java]: it has nothing to with 'UNION'. this method just does kind of a deep-copy of symols. Please rename it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+        {
+            ImmutableMap.Builder<Symbol, Signature> functionsBuilder = ImmutableMap.builder();
+            ImmutableList.Builder<Symbol> groupByKeys = ImmutableList.builder();
+
+            List<Symbol> intersectOutputSymbols = node.getOutputSymbols();
+            Signature countSignature = new Signature("count", AGGREGATE, "bigint", "boolean");
+
+            for (int i = intersectOutputSymbols.size(); i < outputSymbols.size(); i++) {
+                Symbol countSymbol = allocateNewSymbolFrom("count", BIGINT);
+                aggregationsBuilder.put(countSymbol, new FunctionCall(QualifiedName.of("count"), ImmutableList.<Expression>of(new QualifiedNameReference(outputSymbols.get(i).toQualifiedName()))));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java]: what is an argument (which symbol is used) for count()?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+            for (Symbol intersectSymbol : sourceSymbolMap.keySet()) {
+                Symbol symbol = allocateNewSymbolFrom(intersectSymbol);
+                projectionAssignments.put(symbol, sourceSymbolMap.get(intersectSymbol));
+            }
+
+            // add extra marker fields to the projection
+            List<Expression> markers = new ArrayList<>(Collections.nCopies(sources.size(), new Cast(new NullLiteral(), StandardTypes.BOOLEAN)));
+            markers.set(sourceIndex, TRUE_LITERAL);
+            for (Expression markerExpression : markers) {
+                projectionAssignments.put(allocateNewSymbolFrom(INTERSECT_MARKER, BOOLEAN), markerExpression);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:216]: it would be nice to have in a symbol name to which source this marker relates. So in a `EXPLAIN` user could understand what is going on.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestImplementIntersectAsUnion.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.sql.tree.QualifiedName.of;
+import static org.testng.Assert.assertEquals;
+
+public class TestImplementIntersectAsUnion
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestImplementIntersectAsUnion.java]: this test is hard to be maintained. I would remove it. See that no other optimizers have such unit tests. The idea is to have end-to-end query execution. When queries with INTERSECT are working then your optimizer is tested.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+
+import static com.facebook.presto.tests.TestGroups.INTERSECT;
+import static com.facebook.presto.tests.TestGroups.QUERY_ENGINE;
+import static com.teradata.tempto.assertions.QueryAssert.Row.row;
+import static com.teradata.tempto.assertions.QueryAssert.assertThat;
+import static com.teradata.tempto.query.QueryExecutor.query;
+import static com.teradata.tempto.query.QueryType.SELECT;
+
+@Requires(ImmutableNationTable.class)
+public class IntersectTests
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java]: put all these queries to AbstractQueriesTest. Leave one smoke query as product tests.
AbstractQueriesTest are used for all the connectors, not only hive.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+        extends ProductTest
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java]: check EXPLAIN with INTERSECT

Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java]: add test for tables with duplicated rows. And that query returns duplicated rows.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java (hunk 22 above)
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java]: use convention tests for such testing

File: presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    public void intersectWithNonDistinctResults()
+    {
+        assertThat(query("select n_regionkey from nation where n_nationkey < 7 INTERSECT select n_regionkey from nation where n_nationkey > 21"))
+                .containsOnly(row(1), row(3));
+    }
+
+    @Test(groups = {INTERSECT, QUERY_ENGINE})
+    public void intersectWithExplicitDistinctClause()
+    {
+        assertThat(query("select n_regionkey from nation where n_nationkey < 7 INTERSECT DISTINCT select n_regionkey from nation where n_nationkey > 21"))
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java]: `DISTINCT` does not affect results for this query. Please use a query where `DISTINCT` changes the output.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java (hunk 23 above)
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/IntersectTests.java]: add tests for tables with null column values to see if your rewrite properly handles null values in union and aggragation

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 25]
```diff
@@ -1418,6 +1418,24 @@ public void testGroupingCombinationsDistinct()
...
     @Test
+    public void testIntersect()
+            throws Exception
+    {
+        MaterializedResult actual = computeActual("SELECT * FROM (VALUES 1, 2, 3, 4) INTERSECT SELECT * FROM (VALUES 3, 4) INTERSECT SELECT * FROM (VALUES 4)");
+        MaterializedResult expected = resultBuilder(getSession(), BIGINT, BIGINT)
+                .row(4)
+                .build();
+
+        assertEquals(actual.getMaterializedRows(), expected.getMaterializedRows());
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: does the other database not support `INTERSECT`?

you can rewrite this as:
`
assertEquals(query, "values 4");

**Comment by kokosing** [comment]
some comments. Other than that, good work! :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 17 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:77]: In that case, `foo_cnt` will be >1 and `bar_cnt` will be 1 and we are checking the predicate for both, right? Also note that the grouped by is on the UNION ALL output of both tables not just `a` from `foo`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 18 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java]: done. renamed to generateOutputSymbolsFrom

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 20 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:216]: What do you mean by source? I can add the `id` of the source node to the symbol name. But `EXPLAIN` does not show the id of the nodes, so  it wont be obvious which source the marker belongs to.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestImplementIntersectAsUnion.java (hunk 21 above)
Comment (anusudarsan) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestImplementIntersectAsUnion.java]: there are unit tests for other optimizers. But probably those are not checking a node tree like this one. I agree it is kind of hard to maintain. I can remove it

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 17 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:77]: Lets take a look at this example:

```
SELECT a FROM (
  SELECT a, count(F) as FC, count(B) AS BC) FROM (
    SELECT a, true as F, null as B FROM (VALUES 1, 1) t(a) 
    UNION ALL
    SELECT a, null as F, true as B FROM (VALUES 1, 1) t(a) 
  ) T1 GROUP BY A
) T2 where FC >1 AND BC > 1
```

evolves to

```
SELECT a FROM (
  SELECT a, count(F) as FC, count(B) AS BC) FROM (
    VALUES (1, true, null), (1, true, null), (1, null, true), (1, null, true)
  ) T1 GROUP BY A
) T2 where FC >1 AND BC > 1
```

And now grouping to:

```
SELECT a FROM (VALUES (1, 2, 2)) T(a, FC, FB)
) T2 where FC >1 AND BC > 1
```

And as a result of INTERSECT you will get one row `1`, while I would expect two rows `(1), (1)`. Am I correct?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 20 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:216]: yeah. You are right. Please nevermind the comment :)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestImplementIntersectAsUnion.java (hunk 21 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestImplementIntersectAsUnion.java]: I had something like that in flight: https://github.com/prestodb/presto/pull/4404. This could make it easier for you to write unit tests. But this didn't get to master yet. ; /

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 26]
```diff
@@ -777,7 +775,7 @@ public SetOperationPlanner(SetOperation node, Void context)
             return outputSampleWeight;
         }
 
-        public SetOperationPlanner process()
+        public SetOperationPlanner process(Void context)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: do you need any parameter at all? You can just pass `null` to `processAndCoerceIfNecessary`

File: presto-product-tests/src/main/resources/sql-tests/testcases/intersect/intersectAndUnion.result [hunk 27]
```diff
@@ -0,0 +1,7 @@
+-- delimiter: |; ignoreOrder: true;
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/sql-tests/testcases/intersect/intersectAndUnion.result]: If you want to you can group all these 4 convention files into one. It can be done in the same way as: https://github.com/prestodb/tempto/blob/master/tempto-examples/src/main/resources/sql-tests/testcases/nation/allRows.sql

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 28]
```diff
@@ -1421,12 +1421,25 @@ public void testGroupingCombinationsDistinct()
     public void testIntersect()
             throws Exception
     {
-        MaterializedResult actual = computeActual("SELECT * FROM (VALUES 1, 2, 3, 4) INTERSECT SELECT * FROM (VALUES 3, 4) INTERSECT SELECT * FROM (VALUES 4)");
-        MaterializedResult expected = resultBuilder(getSession(), BIGINT, BIGINT)
-                .row(4)
-                .build();
+        assertQuery("select regionkey from nation where nationkey < 7 INTERSECT select regionkey from nation where nationkey > 21", "VALUES 1, 3");
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: can you please format this queries?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 29]
```diff
@@ -1421,12 +1421,25 @@ public void testGroupingCombinationsDistinct()
...
+    @Test
+    public void testIntersectWithAggregation()
+            throws Exception
+    {
+        assertQuery("SELECT count(*) FROM nation INTERSECT SELECT COUNT(regionkey) FROM nation HAVING SUM(regionkey) IS NOT NULL", "SELECT 25");
+        assertQuery("SELECT SUM(nationkey), COUNT(name) FROM (SELECT nationkey,name FROM nation INTERSECT SELECT regionkey,name FROM nation)n", "VALUES (5, 3)");
+        assertQuery("SELECT count(*) * 2 FROM nation INTERSECT (SELECT sum(nationkey) FROM nation GROUP BY regionkey order by 1 limit 2)", "SELECT 50");
+        assertQuery("SELECT COUNT(a) FROM (SELECT nationkey AS a FROM (SELECT nationkey FROM nation INTERSECT SELECT regionkey FROM nation)n1 INTERSECT SELECT regionkey FROM nation) n2", "SELECT 5");
+        assertQuery("select count(*),sum(2), regionkey from (select nationkey, regionkey from nation INTERSECT select regionkey, regionkey from nation)n group by regionkey", "VALUES (1, 2, 0), (1, 2, 1), (1, 2, 4)");
     }
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: add tests for INTERSECT with nested UNION and UNION with nested INTERSECT

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 17 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:77]: It looks like I was wrong: https://msdn.microsoft.com/en-us/library/ms188055.aspx
INTERSECT returns distinct rows always. So is there any purpose for `INTERSECT DISTINCT`?

**Comment by kokosing** [comment]
just few more comments and I am think you are ready to create a pull request in prestodb/presto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java (hunk 17 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java:77]: yes. `INTERSECT` returns distinct rows. The example you mentioned above is `INTERSECT ALL`. `INTERSECT DISTINCT` has the same meaning as `INTERSECT`. Presto also has `UNION DISTINCT` which has same meaning as `UNION`
